Automattic/newspack-popups
# Set up segmentation features

### All Submissions:






### Changes proposed in this Pull Request:



Sets up client visits tracking to enable campaign segmentation. For starters, adding a minimum posts read count option.

This is ready for review, but read on for some high-level questions which might influence large changes.

### How to test the changes in this Pull Request:

1. Create a new campaign, observe a new "Campaign Segmentation" panel in the sidebar
2. Set the "Min. no. of posts read until displayed" setting to 3
3. Visit the site as a non-logged in user, observe that the campaign is displayed only upon the third article reading

### Questions

#### 1. Data stored in the DB

There are now three tables with which this plugin interacts – options, and the two custom tables set up in this PR. A single client visit (assuming there are campaigns on the page) will write to four rows (at most):

1. `<client-id>-popups` transient in options table – to store data about client's interaction with campaigns in general (e.g. has dismissed _a_ newsletter campaign)
2. `<client-id>-<campaign-id>-popup` transient in options table – to store information about a client's interaction with a specific campaign (e.g. last seen date)
3. a row in the clients custom table – the data about the client, which will serve later for segmentation, e.g. information that a client is a newsletter subscriber
4. a row in the visits custom table – data about a single visit

I was thinking about two ways to make it all lighter on the DB:

1. Move the transients data to the custom clients table, making them two additional columns on a client
2. Something suggested by @jeffersonrabb – using just **one custom table**, with visits, donations, newsletter subscriptions, single popup-related data – all as rows.  
One issue with this approach is the mapping of client IDs to user IDs – reconciliation of data from different clients. This is a different topic, so I won't get into details here, but I think that this mapping could be maintained on the WP side – storing the client ID(s) as user meta. The API would have to query the users table on a logged in user's visit, but maybe the cost of that is not an issue.
The "client meta" information, like "has dismissed a newsletter popup" information could be queried form this table, but I don't know what's the performance cost here (querying for all newsletter campaigns seen by a client to determine if at least one was dismissed vs. reading just one "client meta" row)
Update: this approach is pursued in #233

#### 2. DB config

The custom tables configuration – do the column types (which determine the size) make sense? 

#### 3. Should this be an opt-in feature?

The visits-reporting will require a lot of data to be stored, and some users may not need segmentation. But maybe we should revisit that question later if such problems arise.

### Other information:
